File: scripts/nav_bridge.py
#!/usr/bin/env python2
import sys
sys.path.insert(0, '/usr/local/lib/python2.7/dist-packages/perls/lcmtypes')
import lcm
import math
import tf
from acfrlcm import auv_acfr_nav_t
import rospy
from tf.transformations import quaternion_from_euler
from sensor_msgs.msg import NavSatFix
from sensor_msgs.msg import NavSatStatus
from nav_msgs.msg import Odometry
from std_msgs.msg import Header
from sensor_msgs.msg import Imu
import numpy as np

class NavBridge:
    def __init__(self):
        self.lc = lcm.LCM()
        self.lc_sub = self.lc.subscribe("WAMV.ACFR_NAV", self.handler)

        self.fix_pub_ = rospy.Publisher('fix', NavSatFix, queue_size=10)
        self.odom_pub_ = rospy.Publisher('odom',Odometry,queue_size=10)
        self.imu_pub_ = rospy.Publisher('imu/data',Imu,queue_size=10)
        self.use_odom_ = True
        self.use_fix_ = True
        self.use_imu_ = True

        self.gps_frame_ = "novatel"

        self.seq_ = 0

    def unsub(self):
        self.lc.unsubscribe(self.lc_sub)

    def handler(self, channel, data):
        """The LCM callback. Receives the LCM messages, publishes ROS messages.

        Args:
            channel (str): The channel it is received on.
            data (type): Description of parameter `data`.

        Returns:
            type: Description of returned object.

        """
        # Header Message
        headermsg = Header()
        headermsg.seq = self.seq +1
        headermsg.stamp = rospy.Time.now()
        headermsg.frame_id = self.gps_frame_

        # GPS
        msg = auv_acfr_nav_t.decode(data)
        if self.use_fix_ == True:
            rospy.logdebug("Publising gps info")
            fixmsg = NavSatFix()
            fixmsg.header = headermsg
            fixmsg.latitude = math.radians(msg.latitude)
            fixmsg.longitude = math.radians(msg.longitude)
            fixmsg.altitude = msg.altitude
            fixmsg.position_covariance = [0,0,0,0,0,0,0,0,0]
            fixmsg.position_covariance_type = 0
            self.fix_pub_.publish(fixmsg)

        if self.use_imu_ == True:
            rospy.logdebug("Publishing imu")
            imumsg = Imu()
            q = quaternion_from_euler(msg.pitch, msg.roll, -np.unwrap(msg.heading-np.pi/2)) # TODO check this
            imumsg.orientation.x = q[0]
            imumsg.orientation.y = q[1]
            imumsg.orientation.z = q[2]
            imumsg.orientation.w = q[3]

            imumsg.angular_velocity.x = msg.pitchRate
            imumsg.angular_velocity.y = msg.rollRate
            imumsg.angular_velocity.z = -msg.headingRate

            self.imu_pub_.publish(imumsg)

        if self.use_odom_ == True:
            br = tf.TransformBroadcaster()
            q = quaternion_from_euler(msg.pitch, msg.roll, -np.unwrap(msg.heading-np.pi/2)) # TODO check this

            rospy.logdebug("Publishing odom")
            odommsg = Odometry()
            odommsg.header = headermsg
            odommsg.header.frame_id = "odom"

            #TODO, USE TRANSFORM
            odommsg.child_frame_id ="base_link"
            odommsg.pose.pose.position.x = msg.y
            odommsg.pose.pose.position.y = msg.x
            odommsg.pose.pose.position.z = -msg.altitude
            odommsg.pose.pose.orientation.x = q[0]
            odommsg.pose.pose.orientation.y = q[1]
            odommsg.pose.pose.orientation.z = q[2]
            odommsg.pose.pose.orientation.w = q[3]


            odommsg.twist.twist.linear.x = msg.vy
            odommsg.twist.twist.linear.y = msg.vx
            odommsg.twist.twist.linear.z = -msg.vz
            odommsg.twist.twist.angular.x = msg.pitchRate
            odommsg.twist.twist.angular.y = msg.rollRate
            odommsg.twist.twist.angular.z = -msg.headingRate

            self.odom_pub_.publish(odommsg)
            br.sendTransform((msg.y, msg.x, 0),
                         tf.transformations.quaternion_from_euler(msg.pitch, msg.roll, -np.unwrap(msg.heading-np.pi/2)),
                         rospy.Time.now(),
                         "base_link",
                         "odom")

if __name__ == "__main__":
    rospy.init_node('lcmtoRosFix', anonymous=True)

    navbr = NavBridge()

    try:
        while True:
            navbr.lc.handle()
    except KeyboardInterrupt:
        pass

    navbr.unsub()

File: src/nav_bridge.cpp
#include <stdio.h>
#include <ros/ros.h>
#include <lcm/lcm-cpp.hpp>
#include <math.h>
#include "perls-lcmtypes++/acfrlcm/auv_acfr_nav_t.hpp"
#include "perls-lcmtypes++/acfrlcm/auv_relay_t.hpp"
#include <nav_msgs/Odometry.h>
#include <sensor_msgs/NavSatFix.h>
#include <sensor_msgs/Imu.h>
#include <std_msgs/Header.h>
#include <std_msgs/String.h>
#include <tf/transform_datatypes.h>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_broadcaster.h>
#include <geometry_msgs/TransformStamped.h>


class NavBridge
{
    public:
        NavBridge()
        {
            odomPub_ = n_.advertise<nav_msgs::Odometry>("/odom", 1);
            fixPub_ = n_.advertise<sensor_msgs::NavSatFix>("/fix", 1);
            imuPub_ = n_.advertise<sensor_msgs::Imu>("/imu", 1);
            modePub_ = n_.advertise<std_msgs::String>("/vessel_mode", 1);
        }
        ~NavBridge() {}

        void handleMessage(const lcm::ReceiveBuffer* rbuf,
                const std::string& chan,
                const acfrlcm::auv_acfr_nav_t *nav)
        {
            if(ros::ok)
            {
                std_msgs::Header header_msg;
                header_msg.stamp =  ros::Time::now();
                header_msg.seq = seq_;
                seq_++; // Increment the sequence

                tf2::Quaternion quat;
                quat.setRPY(nav->pitch, nav->roll, -NavBridge::unwrap(nav->heading - M_PI_2)); //TODO convert to ENU
                quat.normalize();

                // Fill the Odometry message
                nav_msgs::Odometry odom_msg;
                odom_msg.header = header_msg;
                odom_msg.header.frame_id = "odom";
                odom_msg.child_frame_id ="base_link";
                odom_msg.pose.pose.position.x = nav->y;
                odom_msg.pose.pose.position.y = nav->x;
                odom_msg.pose.pose.position.z = -nav->depth;
                odom_msg.pose.pose.orientation.x = quat[0];
                odom_msg.pose.pose.orientation.y = quat[1];
                odom_msg.pose.pose.orientation.z = quat[2];
                odom_msg.pose.pose.orientation.w = quat[3];

                odom_msg.twist.twist.linear.x = nav->vy;
                odom_msg.twist.twist.linear.y = nav->vx;
                odom_msg.twist.twist.linear.z = -nav->vz;
                odom_msg.twist.twist.angular.x = nav->pitchRate;
                odom_msg.twist.twist.angular.y = nav->rollRate;
                odom_msg.twist.twist.angular.z = -nav->headingRate;

                // Publish the odom message
                odomPub_.publish(odom_msg);

                // IMU
                sensor_msgs::Imu imu_msg;
                imu_msg.orientation.x = quat[0];
                imu_msg.orientation.y = quat[1];
                imu_msg.orientation.z = quat[2];
                imu_msg.orientation.w = quat[3];

                imu_msg.angular_velocity.x = nav->pitchRate;
                imu_msg.angular_velocity.y = nav->rollRate;
                imu_msg.angular_velocity.z = -nav->headingRate;

                imuPub_.publish(imu_msg);

                // Fix Publish
                sensor_msgs::NavSatFix fix_msg;
                fix_msg.header = header_msg;
                fix_msg.header.frame_id = "novatel";
                fix_msg.latitude = nav->latitude;
                fix_msg.longitude = nav->longitude;
                fix_msg.altitude = nav->altitude;
                fixPub_.publish(fix_msg);

                // Send the transform
                static tf2_ros::TransformBroadcaster br;
                geometry_msgs::TransformStamped trStamp;
                trStamp.header = header_msg;
                trStamp.header.frame_id = "odom";
                trStamp.child_frame_id = "base_link";
                trStamp.transform.translation.x = odom_msg.pose.pose.position.x;
                trStamp.transform.translation.y = odom_msg.pose.pose.position.y;
                trStamp.transform.translation.z = odom_msg.pose.pose.position.z;
                trStamp.transform.translation.z = odom_msg.pose.pose.position.z;
                trStamp.transform.rotation.x = quat.x();
                trStamp.transform.rotation.y = quat.y();
                trStamp.transform.rotation.z = quat.z();
                trStamp.transform.rotation.w = quat.w();

                br.sendTransform(trStamp);
            }

        }
        // Keeps the angle between -pi, pi
        double unwrap(double x)
        {
            x = fmod(x + M_PI,2*M_PI);
            if (x < 0)
                x += 2*M_PI;
            return x - M_PI;
        }


        void handleMode(const lcm::ReceiveBuffer* rbuf,
                const std::string& chan,
                const acfrlcm::auv_relay_t *msg)
        {
            if(ros::ok)
            {
                std::string mode_colour;
                if(msg->state == 0)
                {
                    mode_colour = "red";
                }
                else if(msg->state == 1)
                {
                    mode_colour = "orange";
                }
                else if(msg->state == 2)
                {
                    mode_colour = "green";
                }
                std_msgs::String mode_msg;
                mode_msg.data = mode_colour;
                modePub_.publish(mode_msg);
            }
        }
         
    private:
        ros::NodeHandle n_;
        ros::Publisher odomPub_;
        ros::Publisher imuPub_;
        ros::Publisher fixPub_;
        ros::Publisher modePub_;
        int seq_;

};

int main(int argc, char** argv)
{
    lcm::LCM lcm;

    ros::init(argc, argv, "nav_bridge");

    if(!lcm.good())
        return 1;

    NavBridge nb;
    lcm.subscribe("WAMV.ACFR_NAV", &NavBridge::handleMessage, &nb);
    lcm.subscribe("WAMV.ACFR_NAV", &NavBridge::handleMode, &nb);

    while(0 == lcm.handle());

    return 0;
}
